### ISSUE TITLE
end `simplify_network()` immediately if already meets criteria

### ISSUE THREAD

**Issue by jGaboardi** [uscuni/neatnet#68]
We need some logic to end `sgeop.simplify_network()` immediately if the input roads dataset already meets our simplififed criteria. This can be determined within the call to [`sgeop.nodes.fix_topology()`](https://github.com/uscuni/sgeop/blob/681bc5e7d3403098a884dac7d399551e7e6f6a1d/sgeop/simplify.py#L507) and simply return the result there.

The check should be after [`induce_nodes()`](https://github.com/uscuni/sgeop/blob/681bc5e7d3403098a884dac7d399551e7e6f6a1d/sgeop/nodes.py#L276) is called where we are checking if `roads` and `roads_w_nodes` are equivalent. If they are, no further simplification is needed, as far as I understand.

Thoughts @martinfleis @anastassiavybornova ?

**Comment by martinfleis** [comment]
> If they are, no further simplification is needed, as far as I understand

wrong. Topological correctness does not mean that there are no artifacts.

**Comment by jGaboardi** [comment]
So that would be [here](https://github.com/uscuni/sgeop/blob/681bc5e7d3403098a884dac7d399551e7e6f6a1d/sgeop/simplify.py#L512) after the first loop is completed, correct? (if that `artifacts` df is empty)

**Comment by jGaboardi** [comment]
When [this condition is met](https://github.com/uscuni/sgeop/blob/50fc60576cae0c323309454296213d45cb03d3b2/sgeop/artifacts.py#L52-L55), should it raise an error, or can we actually the input simplified enough and return? It seems like this could be true on both the first and second loops?

**Comment by martinfleis** [comment]
> or can we actually the input simplified enough and return

We can't. Threshold detection occasionally fails in certain contexts due to the distribution of block shapes. It should raise as it does now.

**Comment by jGaboardi** [comment]
OK, I'm wondering if it would be prudent to cut a patch release of `momepy` after https://github.com/pysal/momepy/pull/666 so at least we can fail gracefully when an skeleton is passed in that can't be polygonized?

**Comment by jGaboardi** [comment]
After that I can look back to try to determine better logic this.

**Comment by martinfleis** [comment]
We'd need to backport that. But if you wait a bit, I think that Streetscape will be ready for 0.9.

**Comment by jGaboardi** [comment]
Yeah, I think this can surely wait. Just wanted to keep it on the radar.

**Comment by jGaboardi** [comment]
So I have been playing around with some ideas and thinking on this, and (unless I am missing something obvious) the logic gets a bit subjective. We may simply want to make it clear to the user that the algorithm we employ does not decide if a network dataset is "simplified enough" and the user themself needs to determine that.

As a case in point, below is some code and plots of the original Apalachicola, FL testing data (`SI-0`) and three successive rounds of simplification (`SI-{1,2,3}`). `SI-{1,2}` are both clean results, with `SI-2` being slightly over-simplified. However, `SI-3` leads to dramatic over-simplification while not being "wrong." It should be noted that a fourth round of simplification fails with [our current logic](https://github.com/uscuni/sgeop/blob/b5db0feb007f213438b31af218f574f8a6285b1b/sgeop/artifacts.py#L96-L100)

My goal with this was to find some simple & clear criterion for popping out of the process, but this is seems to be more difficult than I had imagined These are some relevant changes on the `@GH68_early_terminate_if_all_good` branch in:
* [`simplify.simplify_network()`](https://github.com/uscuni/sgeop/blob/0a605992b86e98ddfedf260f25bddc2308af4ff7/sgeop/simplify.py#L593-L622)
* [`artifacts.get_artifacts()`](https://github.com/uscuni/sgeop/blob/0a605992b86e98ddfedf260f25bddc2308af4ff7/sgeop/artifacts.py#L90-L94)

```python
import geopandas, matplotlib.pyplot, sgeop

f = "sgeop/tests/data/apalachicola_original.parquet"
raw_original = geopandas.read_parquet(f, columns=["geometry"]).reset_index(drop=True)

iterations = 3
scene = "Simplification Iteration {}"
scenes = {scene.format(0): raw_original}

for i in range(iterations):
    scenes[scene.format(i+1)] = sgeop.simplify_network(scenes[scene.format(i)].copy())

panels = [[scene.format(0), scene.format(1)], [scene.format(2), scene.format(3)]]
fig, axd = matplotlib.pyplot.subplot_mosaic(panels, figsize=(9, 9), layout="constrained")
for k, ax in axd.items():
    scenes[k].plot(ax=ax)
    ax.axis("off")
    ax.set_title(k)
```

Perhaps we should leave this issue open for now, but I'll not spend more time on it at this point.

------------------------------------------------------

![simplification_iterations](https://github.com/user-attachments/assets/99f7088f-7682-40ea-b94f-d5ab8530228d)

**Comment by martinfleis** [comment]
One thing about the subsequent runs is that those should use the threshold identified in the original data. Otherwise the artifact detection is broken, hence the result of the last two iterations. If you set it to 7 or something like that, it will stabilise and stop making changes eventually.

**Comment by jGaboardi** [comment]
Sounds good. Closing for now. We can reopen & revisit in the future if it comes up again.

**Comment by jGaboardi** [comment]
Leaving the [@GH68_early_terminate_if_all_good](https://github.com/uscuni/sgeop/tree/GH68_early_terminate_if_all_good) branch intact for now.

**Comment by martinfleis** [comment]
I failed to figure out a logic that would automatically stop the loop. In Liege, there are some minor changes even in loop 88 (that is when I lost patience) but there's no point doing that. It may be that one intersection is just being redrawn in a slightly different way.

What we can do is exposing the keyword for number of loops to do, defaulting to 2. Below is the total area of artifacts at each loop.

```
entering loop 2 367806.21306513844
entering loop 3 324077.39595179004
entering loop 4 300499.03915179556
entering loop 5 290481.8624264282
entering loop 6 285826.67560476495
entering loop 7 281478.3192801817
entering loop 8 280790.6230465451
entering loop 9 280605.62870734185
entering loop 10 280465.19804152835
entering loop 11 280261.5668731042
entering loop 12 280107.3708299049
entering loop 13 279958.34173831576
entering loop 14 279614.6234327308
entering loop 15 279594.8716519449
entering loop 16 279383.0226339551
entering loop 17 279262.5297435447
entering loop 18 304528.7285142959
entering loop 19 299242.9661496359
entering loop 20 296864.43352820486
entering loop 21 296196.657502414
entering loop 22 295051.3553276628
entering loop 23 294052.8613620213
entering loop 24 293075.15948068956
entering loop 25 292866.96485431824
entering loop 26 293019.91153932165
entering loop 27 292948.35589333845
entering loop 28 292763.8212925927
entering loop 29 293257.47410003375
entering loop 30 293128.681893803
entering loop 31 293379.10329105105
entering loop 32 293188.2913709269
entering loop 33 292923.5796961089
entering loop 34 292743.2136307505
entering loop 35 292958.8097953589
entering loop 36 292753.4355825776
entering loop 37 292470.68499542074
entering loop 38 292361.65520792705
entering loop 39 292501.974145582
entering loop 40 293492.8513721544
entering loop 41 293310.87109786284
entering loop 42 293130.80854489835
entering loop 43 293003.7937956833
entering loop 44 293178.2216976078
entering loop 45 292942.71969571826
entering loop 46 292717.9924517356
entering loop 47 292595.5667462898
entering loop 48 287788.87071702676
entering loop 49 287725.6249939173
entering loop 50 287658.06089957984
entering loop 51 287591.77566735615
entering loop 52 287454.6566106929
entering loop 53 287338.79137938167
entering loop 54 287242.0347660833
entering loop 55 287157.3844407047
entering loop 56 287003.5467430129
entering loop 57 286953.25934665964
entering loop 58 286836.2946799854
entering loop 59 287800.52914685616
entering loop 60 287731.56008665543
entering loop 61 287663.80366017803
entering loop 62 287597.33183043177
entering loop 63 287462.47599289485
entering loop 64 287347.1276780327
entering loop 65 287250.67301085906
entering loop 66 287182.17897742323
entering loop 67 287788.8702160654
entering loop 68 287725.62479437154
entering loop 69 287658.05550924334
entering loop 70 287591.7768559301
entering loop 71 287454.6603369246
entering loop 72 287338.7907737046
entering loop 73 287242.03662480495
entering loop 74 287157.381891743
entering loop 75 287003.55105508416
entering loop 76 286953.2609870756
entering loop 77 286836.29530392226
entering loop 78 287800.5261323042
entering loop 79 287731.55685352394
entering loop 80 287663.8056873989
entering loop 81 287597.3411936541
entering loop 82 287462.4881665685
entering loop 83 287347.1323668562
entering loop 84 287250.67806838633
entering loop 85 287182.1862473519
entering loop 86 287788.87831666565
entering loop 87 287725.6364584869
entering loop 88 287658.0653025046
```

**Comment by martinfleis** [comment]
Here is the file and here are some of the changes.

[loop.gpkg.zip](https://github.com/user-attachments/files/19217658/loop.gpkg.zip)

<img width="762" alt="Image" src="https://github.com/user-attachments/assets/ed4fcdfa-7c8b-4608-a2aa-21fd9b9d8d63" />

What happens is that there are places where we can really eliminate the artefact as it contains a non-planar edge, it is just progressively getting narrower and weirder. As a consequence, the network in loop 7 is actually worse than in the loop 2.  

<img width="755" alt="Image" src="https://github.com/user-attachments/assets/8f922a96-445f-45ec-842c-7bcf036f1dd6" />

My proposal is to drop this. If reviewers pick this up, it is an easy thing to respond to. Otherwise, there's probably no point going beyond two loops.

**Comment by martinfleis** [comment]
And here is the branch if you want to play with it https://github.com/uscuni/neatnet/tree/loooooooop

**Comment by jGaboardi** [comment]
"looooooooooooooop"

**Comment by jGaboardi** [comment]
Great work up. So basically you're proposing a "do nothing for now" on our side, right? I agree with this proposal.

**Comment by martinfleis** [comment]
I would probably expose that keyword and suggest in the docstring to keep it default :).
